adr/0023-multi-node-kubernetes.md

```diff
@@ -0,0 +1,75 @@
+# 22. Multi-Node Kubernetes with `zarf init`
+
+Date: 2024-01-23
+
+## Status
+
+Pending
+
+## Context
+
+Currently today, when allowing `zarf init` to handle cluster creation, Zarf doesn't have ability to automatically or semi-automatically provision itself across multiple nodes.  The idea here would be to allow for horizontal scalability across multiple virtual or physical nodes for site reliability and automatic failover.
+
+The first pass would consider scaling horizontally with the Embedded DB model.  There would be minimal changes on the current k3s.service config.  The change required here would be to include a shared token.  By default, if K3S doesn't receive a token it will auto generate one.  You can reset this on an existing cluster by running `k3s token rotate --new-token=foo`.
+
+If one wanted to specify a token in advance they could simply modify their existing `zarf deploy/init`, `--set K3S_ARGS` command to include `--token=foo`
+
+For example:
+```shell
+zarf init --components=k3s,git-server --confirm --set K3S_ARGS=\"--disable=traefik --disable=metrics-server --disable=servicelb --tls-san=1.2.3.4 --token=foo\""
+```
+
+This results in a line as such for example:
+
+```ini
+ExecStart=/usr/sbin/k3s server --write-kubeconfig-mode=700 --write-kubeconfig /root/.kube/config --disable=traefik --disable=metrics-server --disable=servicelb --tls-san=1.2.3.4 --token=foo
```

Comment:
I believe we will need to use the `cluster-init` flag to enable clustering in HA embedded etcd mode:

> To get started, first launch a server node with the cluster-init flag to enable clustering and a token that will be used as a shared secret to join additional servers to the cluster.

https://docs.k3s.io/datastore/ha-embedded
https://docs.k3s.io/cli/server#cluster-options

This is what tells k3s to use the embedded etcd datastore rather than the default SQLite datastore

Reply:
Yes you are right, I wrote this before testing out multiple `master` nodes.